### PR TITLE
Header layering proposal.

### PR DESCRIPTION
**Pull request by davidchisnall**
Straw-man header layering, mostly based on prior discussions with some
tweaks from actually trying to do it.

See src/snmalloc/README.md for an explanation of the layers.

### CODE REVIEW

File: src/test/func/jemalloc/jemalloc.cc [hunk 1]
```diff
@@ -9,8 +9,8 @@
 #define SNMALLOC_BOOTSTRAP_ALLOCATOR
 #define SNMALLOC_JEMALLOC3_EXPERIMENTAL
 #define SNMALLOC_JEMALLOC_NONSTANDARD
-#include "../../../override/jemalloc_compat.cc"
-#include "../../../override/malloc.cc"
+#include "../../../snmalloc/override/jemalloc_compat.cc"
```
Comment (mjp41) [inline src/test/func/jemalloc/jemalloc.cc:12]: Most tests seem to get away with `<snmalloc/...>`

File: src/test/func/malloc/malloc.cc [hunk 2]
```diff
@@ -6,7 +6,7 @@
 #undef SNMALLOC_NO_REALLOCARRAY
 #undef SNMALLOC_NO_REALLOCARR
 #define SNMALLOC_BOOTSTRAP_ALLOCATOR
-#include "../../../override/malloc.cc"
+#include "../../../snmalloc/override/malloc.cc"
```
Comment (mjp41) [inline src/test/func/malloc/malloc.cc:9]: Most tests seem to get away with `<snmalloc/...>`

File: src/test/func/domestication/domestication.cc [hunk 3]
```diff
@@ -98,7 +98,7 @@ namespace snmalloc
 }
 
 #  define SNMALLOC_NAME_MANGLE(a) test_##a
-#  include "../../../override/malloc.cc"
+#  include "../../../snmalloc/override/malloc.cc"
```
Comment (mjp41) [inline src/test/func/domestication/domestication.cc:101]: Most tests use `<snmalloc/...>`

File: src/snmalloc/backend_helpers/pagemap.h [hunk 4]
```diff
@@ -1,9 +1,7 @@
 #pragma once
 
-#include "../ds/bits.h"
-#include "../ds/helpers.h"
+#include "../ds/ds.h"
 #include "../mem/entropy.h"
```
Comment (mjp41) [inline src/snmalloc/backend_helpers/pagemap.h:4]: Should this include `entropy.h` directly?

**Review by mjp41 — APPROVED** [review]
Like the changes.  I think this makes things much cleaner.  Couple of minor nits and CI isn't happy yet.  I can take a second pass when CI approves.

**Review by SchrodingerZhu — APPROVED** [review]

File: src/snmalloc/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+This allows specific files to be moved to a lower layer if appropriate, without too much code churn.
+
+There is only one exception to this rule: `backend/globalconfig.h`.
+This file defines either the default configuration *or* nothing, depending on whether the user has defined `SNMALLOC_PROVIDE_OWN_CONFIG`.
+The layers above the back end should include only this file, so that there is a single interception point for externally defined back ends.
+
+External code should include only the following files:
+
+ - `snmalloc/snmalloc_core.h` includes everything up to `backend_helpers`.
+   This provides the building blocks required to assembly an snmalloc instance, but does not assume any global configuration.
```
Comment (nwf) [inline src/snmalloc/README.md:37]:
```suggestion
   This provides the building blocks required to assemble an snmalloc instance, but does not assume any global configuration.
```

File: src/test/setup.h [hunk 6]
```diff
@@ -1,10 +1,9 @@
 #if defined(SNMALLOC_CI_BUILD)
-#  include <pal/pal.h>
+#  include <snmalloc/pal/pal.h>
 #  if defined(WIN32)
-#    include <ds/bits.h>
 #    include <iostream>
-#    include <pal/pal.h>
 #    include <signal.h>
+#    include <snmalloc/ds/bits.h>
```
[the remark this answers is not in the document]
Comment (davidchisnall) [reply src/test/setup.h:6]: I was largely ignoring the tests for the purpose of layering:  Nothing should depend on the tests, so they can do whatever they want.

**Review by nwf — COMMENTED** [review]
The `ds` directory's new contents confuse me.  I'd have imagined that `aba.h`, `csv.h`, `redblacktree.h`, `seqset.h`, and `singleton.h` were suitable for `ds_core` (at a glance, I don't see why they're not?).  `flaglock.h` might plausibly depend on AAL or PAL, but at present it does not seem to.  `mpmcstack.h` depends on `allocconfig.h` only for `CACHELINE_SIZE`, which seems like it could be templated out instead (and, while here, perhaps we should stop this file putting `SNMALLOC_THREAD_SANITIZER_ENABLED` in the namespace?).

**Comment by davidchisnall** [comment]
> The `ds` directory's new contents confuse me. I'd have imagined that `aba.h`,

Depends on the AAL for `pause`.  We could make it templated on the AAL, but since the AAL is a singleton that seems painful to do unless we are going to need it at lower layers.  Even if it were templated, we wouldn't be able to use it below the AAL layer, so it's only a benefit to have it lower if we want to use it in the PAL.  

> `csv.h`,

Was part of the old pre-`MessageBuilder` stats infrastructure, deleted.

> `redblacktree.h`,

Agreed.

> `seqset.h`,

Agreed.

> and `singleton.h`

Depends on `FlagLock`.

> were suitable for `ds_core` (at a glance, I don't see why they're not?). `flaglock.h` might plausibly depend on AAL or PAL, but at present it does not seem to. 

`FlagLock` depends on the AAL for `pause`.

> `mpmcstack.h` depends on `allocconfig.h` only for `CACHELINE_SIZE`, which seems like it could be templated out instead (and, while here, perhaps we should stop this file putting `SNMALLOC_THREAD_SANITIZER_ENABLED` in the namespace?).

Depends on `ABA`, which depends on the AAL for `pause`.

**Review by mjp41 — APPROVED** [review]
LGTM

**Pull request merged**